staking/integration-tests/tests/stability.rs

```diff
@@ -0,0 +1,833 @@
+        for _ in 0..NUM_OPERATIONS {
+            let operation = get_random_operation(g, &operations, &publishers, &delegators);
+            operations.push(operation);
+        }
+
+        StabilityTestProps {
+            operations,
+            publishers,
+            delegators,
+        }
+    }
+}
+
+type OperationWrapper =
+    Box<dyn Fn(&mut Gen, &[Operation], &[KeypairWrapper], &[KeypairWrapper]) -> Operation>;
+
+struct OperationWeight {
+    weight:    u64,
+    operation: OperationWrapper,
+}
+
+fn get_random_operation(
+    g: &mut Gen,
+    prev_operations: &[Operation],
+    publishers: &[KeypairWrapper],
+    delegators: &[KeypairWrapper],
+) -> Operation {
+    let operations = vec![
+        OperationWeight {
+            weight:    500,
+            operation: Box::new(
+                |g, _operations, publishers, delegators| Operation::Delegate {
+                    publisher: usize::arbitrary(g) % publishers.len(),
+                    delegator: usize::arbitrary(g) % delegators.len(),
+                    amount:    u64::arbitrary(g) % MAX_DELEGATION_AMOUNT + 1,
+                },
+            ),
+        },
+        OperationWeight {
+            weight:    250,
+            operation: Box::new(|g, operations, publishers, delegators| {
+                let del_ops: Vec<&Operation> = operations
+                    .iter()
+                    .filter(|op| matches!(op, Operation::Delegate { .. }))
+                    .collect();
+                let del_op = g.choose(&del_ops);
+                if let Some(Operation::Delegate {
+                    publisher,
+                    delegator,
+                    amount,
+                }) = del_op
+                {
+                    Operation::Undelegate {
+                        publisher: *publisher,
+                        delegator: *delegator,
+                        amount:    min(u64::arbitrary(g) % MAX_UNDELEGATION_AMOUNT + 1, *amount),
+                    }
+                } else {
+                    Operation::Undelegate {
+                        publisher: usize::arbitrary(g) % publishers.len(),
+                        delegator: usize::arbitrary(g) % delegators.len(),
+                        amount:    u64::arbitrary(g) % MAX_UNDELEGATION_AMOUNT + 1,
+                    }
+                }
+            }),
+        },
+        OperationWeight {
+            weight:    250,
+            operation: Box::new(|g, _operations, publishers, delegators| {
+                Operation::AdvanceDelegationRecord {
+                    publisher: usize::arbitrary(g) % publishers.len(),
+                    delegator: usize::arbitrary(g) % delegators.len(),
+                }
+            }),
+        },
+        OperationWeight {
+            weight:    10,
+            operation: Box::new(
+                |g, _operations, _publishers, _delegators| Operation::Advance {
+                    cap: u64::arbitrary(g) % MAX_CAP + 1,
```

Comment:
nit: the +1 means that the maximum is actually MAX_CAP + 1. Maybe it would be clearer if you did like in slash everywhere: `u64::arbitrary(g) % (FRAC_64_MULTIPLIER - 1) + 1,`

Reply:
I think the maximum is actually MAX_CAP because x % y < y.